modules/configuration/src/test/java/org/apache/ignite/internal/configuration/notifications/ConfigurationListenerTest.java

```diff
@@ -644,4 +667,366 @@ public void dataRace() throws Exception {
+            ),
+            events
+        );
+
+        checkListeners(
+            () -> configuration.elements().get("1").elements2().get("2").i().update(200),
+            List.of(
+                "root",
+                "root.elements",
+                "root.elements.onUpd",
+                "root.elements.onUpd",
+                //
+                "root.elements.any",
+                "root.elements.any.elements2",
+                "root.elements.any.elements2.onUpd",
+                "root.elements.any.elements2.any",
+                "root.elements.any.elements2.any.i"
+            ),
+            events
+        );
+
+        checkListeners(
+            () -> configuration.elements().get("1").elements2().change(c -> c.rename("2", "2x")),
+            List.of(
+                "root",
+                "root.elements",
+                "root.elements.onUpd",
+                "root.elements.onUpd",
+                //
+                "root.elements.any",
+                "root.elements.any.elements2",
+                "root.elements.any.elements2.onRen"
+            ),
+            events
+        );
+
+        checkListeners(
+            () -> configuration.elements().get("1").elements2().change(c -> c.delete("2x")),
+            List.of(
+                "root",
+                "root.elements",
+                "root.elements.onUpd",
+                "root.elements.onUpd",
+                //
+                "root.elements.any",
+                "root.elements.any.elements2",
+                "root.elements.any.elements2.onDel",
+                "root.elements.any.elements2.any"
+            ),
+            events
+        );
+    }
+
+    /**
+     * Helper method for testing listeners.
+     *
+     * @param changeFun Configuration change function.
+     * @param exp Expected list of executing listeners.
+     * @param act Reference to the list of executing listeners that is filled after the {@code changeFun} is executed.
+     * @throws Exception If failed.
+     */
+    private static void checkListeners(
+        Supplier<CompletableFuture<Void>> changeFun,
+        List<String> exp,
+        List<String> act
+    ) throws Exception {
+        act.clear();
+
+        changeFun.get().get(1, SECONDS);
+
+        assertEquals(exp, act);
+    }
+
+    /**
+     * @param consumer Consumer of the notification context.
+     * @return Config value change listener.
+     */
+    private static <T> ConfigurationListener<T> configListener(Consumer<ConfigurationNotificationEvent<T>> consumer) {
+        return ctx -> {
+            consumer.accept(ctx);
+
+            return completedFuture(null);
+        };
+    }
+
+    /**
+     * @param consumer Consumer of the notification context.
+     * @return Named config value change listener.
+     */
+    private static <T> ConfigurationNamedListListener<T> configNamedListenerOnUpdate(
+        Consumer<ConfigurationNotificationEvent<T>> consumer
+    ) {
+        return ctx -> {
```

Comment:
Why is ConfigurationNamedListListener suddenly became FunctionalInterface? I think you should provide default implementation for "onUpdate" as well